[FSDP2] Added test to show rank 0 broadcast meta-device flow

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124787
* #124768
* #124767
* #124741
* __->__ #124651

This PR includes two things:
1. Changes to support `load_state_dict(assign=True)`
    - These changes are not ideal, but until we have `DTensor` padding the local tensor and general `swap_tensors` adoption, we may need to make do.
2. Example of how to convert a full state dict on rank 0 to sharded state dict on all ranks via broadcast
    - ~~To-do: check for `recordStream` from the funcol broadcast; if being called, remediate either via `async_op=False` c10d broadcast or use `TORCH_NCCL_AVOID_RECORD_STREAMS=1`~~ switched to using c10d `async_op=False` broadcast
    - To-do: check for broadcast latency since not using any coalescing

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Commit:
[FSDP2] Added test to show rank 0 broadcast meta-device flow

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added test to show rank 0 broadcast meta-device flow"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -161,6 +161,13 @@ def __init__(
         self._init_extensions()
         self.all_gather_outputs: List[torch.Tensor] = []
         self._param_fqn: Optional[str] = None  # prefixed from root module
+        # TODO: Remove this padding logic once DTensor pads the local tensor:
+        # https://github.com/pytorch/pytorch/issues/113045
+        self._post_load_hook_handle = (
+            module_info.module.register_load_state_dict_post_hook(
+                lambda *args, **kwargs: self.reset_sharded_param()
+            )
+        )
```
This is needed to support `load_state_dict(assign=True)` today.

In the future, we want (1) `DTensor` to pad its local tensor and (2) for users to use the `swap_tensors` path for `load_state_dict()` with FSDP2. Together these should allow us to remove this hook.

Commit:
Update on "[FSDP2] Added test to show rank 0 broadcast meta-device flow"


This PR includes two things:
1. Changes to support `load_state_dict(assign=True)`
    - These changes are not ideal, but until we have `DTensor` padding the local tensor and general `swap_tensors` adoption, we may need to make do.
2. Example of how to convert a full state dict on rank 0 to sharded state dict on all ranks via broadcast
    - To-do: check for `recordStream` from the funcol broadcast; if being called, remediate either via `async_op=False` c10d broadcast or use `TORCH_NCCL_AVOID_RECORD_STREAMS=1`
    - To-do: check for broadcast latency since not using any coalescing

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_init.py:
```diff
@@ -554,6 +566,91 @@ def test_invalid_meta_device_init(self):
+        # broadcasting and sharding
+        meta_sharded_sd = model.state_dict()
+        sharded_sd = {}
+        if self.rank == 0:
```
If per-tensor `broadcast` is achieving poor network utilization and is intolerable, then the user/author of this converter can change this to do bucketing with `dist._broadcast_coalesced`. The implementation will be more involved though.

wanchaol replied:
In future I think we can also add a dedicated API maybe: `distribute_multi_tensor` that shard a list of tensors with the same placements

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_init.py:
```diff
@@ -554,6 +566,91 @@ def test_invalid_meta_device_init(self):
+
+        # Construct a sharded state dict from the rank 0 full state dict by
+        # broadcasting and sharding
+        meta_sharded_sd = model.state_dict()
```
The idea behind this conversion is that the since `DTensor`s encode the sharding info at the tensor level, we can use the sharded meta-device state dict to know how to shard the unsharded tensors from the full state dict (via `distribute_tensor()`).

Commit:
Update on "[FSDP2] Added test to show rank 0 broadcast meta-device flow"


This PR includes two things:
1. Changes to support `load_state_dict(assign=True)`
    - These changes are not ideal, but until we have `DTensor` padding the local tensor and general `swap_tensors` adoption, we may need to make do.
2. Example of how to convert a full state dict on rank 0 to sharded state dict on all ranks via broadcast
    - ~~To-do: check for `recordStream` from the funcol broadcast; if being called, remediate either via `async_op=False` c10d broadcast or use `TORCH_NCCL_AVOID_RECORD_STREAMS=1`~~ switched to using c10d `async_op=False` broadcast
    - To-do: check for broadcast latency since not using any coalescing

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added test to show rank 0 broadcast meta-device flow"


This PR includes two things:
1. Changes to support `load_state_dict(assign=True)`
    - These changes are not ideal, but until we have `DTensor` padding the local tensor and general `swap_tensors` adoption, we may need to make do.
2. Example of how to convert a full state dict on rank 0 to sharded state dict on all ranks via broadcast
    - ~~To-do: check for `recordStream` from the funcol broadcast; if being called, remediate either via `async_op=False` c10d broadcast or use `TORCH_NCCL_AVOID_RECORD_STREAMS=1`~~ switched to using c10d `async_op=False` broadcast
    - To-do: check for broadcast latency since not using any coalescing

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added test to show rank 0 broadcast meta-device flow"


This PR includes two things:
1. Changes to support `load_state_dict(assign=True)`
    - These changes are not ideal, but until we have `DTensor` padding the local tensor and general `swap_tensors` adoption, we may need to make do.
2. Example of how to convert a full state dict on rank 0 to sharded state dict on all ranks via broadcast
    - ~~To-do: check for `recordStream` from the funcol broadcast; if being called, remediate either via `async_op=False` c10d broadcast or use `TORCH_NCCL_AVOID_RECORD_STREAMS=1`~~ switched to using c10d `async_op=False` broadcast
    - To-do: check for broadcast latency since not using any coalescing

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:
```diff
@@ -148,9 +148,14 @@ def _test_state_dict_save_load(self, model: nn.Module):
                 param.to_local(),
                 torch.ones_like(param.to_local()) * new_fill_value,
             )
-            self.assertEqual(
```
The assertion was too strict before, and our implementation was actually incorrect since we did not re-pad the local tensor.

Commit:
Update on "[FSDP2] Added test to show rank 0 broadcast meta-device flow"


This PR includes two things:
1. Changes to support `load_state_dict(assign=True)`
    - These changes are not ideal, but until we have `DTensor` padding the local tensor and general `swap_tensors` adoption, we may need to make do.
2. Example of how to convert a full state dict on rank 0 to sharded state dict on all ranks via broadcast
    - ~~To-do: check for `recordStream` from the funcol broadcast; if being called, remediate either via `async_op=False` c10d broadcast or use `TORCH_NCCL_AVOID_RECORD_STREAMS=1`~~ switched to using c10d `async_op=False` broadcast
    - To-do: check for broadcast latency since not using any coalescing

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm!

awgu commented:
@pytorchbot merge

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_init.py:
```diff
@@ -554,6 +566,91 @@ def test_invalid_meta_device_init(self):
+                sharded_tensor = distribute_tensor(
+                    full_param, mesh, sharded_meta_param.placements
+                )
+                sharded_sd[param_name] = nn.Parameter(sharded_tensor)
```
curious if buffers are sometimes included in state dict? for llama2, the buffer is RotaryPositionalEmbeddings.theta but it's not included in HF state dict
